File: dertetgpt-ios/DertetHarness/Network/ModelsAPI.swift
import Foundation

enum ModelsAPI {
    static func fetchAvailableModels(style: ApiStyle, baseUrl: String, apiKey: String) async -> [String] {
        do {
            switch style {
            case .openAICompatible: return try await fetchOpenAIStyle(baseUrl: baseUrl, apiKey: apiKey)
            case .anthropic: return try await fetchAnthropic(baseUrl: baseUrl, apiKey: apiKey)
            case .gemini: return try await fetchGemini(baseUrl: baseUrl, apiKey: apiKey)
            }
        } catch {
            return []
        }
    }

    private static func fetchOpenAIStyle(baseUrl: String, apiKey: String) async throws -> [String] {
        guard !baseUrl.isEmpty, let url = URL(string: baseUrl.trimmedTrailingSlash + "/models") else { return [] }
        var request = URLRequest(url: url)
        if !apiKey.isEmpty { request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization") }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { return [] }
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let list = asArray(obj["data"]) else { return [] }
        return list.compactMap { asString(asDict($0)?["id"]) }.sorted()
    }

    private static func fetchAnthropic(baseUrl: String, apiKey: String) async throws -> [String] {
        guard !apiKey.isEmpty, let url = URL(string: baseUrl.trimmedTrailingSlash + "/models?limit=1000") else { return [] }
        var request = URLRequest(url: url)
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { return [] }
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let list = asArray(obj["data"]) else { return [] }
        return list.compactMap { asString(asDict($0)?["id"]) }.sorted()
    }

    private static func fetchGemini(baseUrl: String, apiKey: String) async throws -> [String] {
        guard !apiKey.isEmpty, let url = URL(string: baseUrl.trimmedTrailingSlash + "/models?pageSize=1000") else { return [] }
        var request = URLRequest(url: url)
        request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { return [] }
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let list = asArray(obj["models"]) else { return [] }
        return list.compactMap { entry -> String? in
            guard let dict = asDict(entry), let name = asString(dict["name"]) else { return nil }
            let methods = asArray(dict["supportedGenerationMethods"])?.compactMap { $0 as? String } ?? []
            let supportsChat = methods.isEmpty || methods.contains("generateContent") || methods.contains("streamGenerateContent")
            guard supportsChat else { return nil }
            return name.hasPrefix("models/") ? String(name.dropFirst("models/".count)) : name
        }.sorted()
    }
}

File: dertet-harness-desktop/agent/llmClients/modelsApi.ts
import { ApiStyle } from "../types";

async function fetchOpenAiStyle(baseUrl: string, apiKey: string): Promise<string[]> {
  if (!baseUrl) return [];
  const headers: Record<string, string> = {};
  if (apiKey) headers.Authorization = `Bearer ${apiKey}`;
  const res = await fetch(`${baseUrl.replace(/\/$/, "")}/models`, { headers });
  if (!res.ok) return [];
  const json: any = await res.json();
  const list = json?.data;
  if (!Array.isArray(list)) return [];
  return list.map((m: any) => m.id).filter(Boolean).sort();
}

async function fetchAnthropic(baseUrl: string, apiKey: string): Promise<string[]> {
  if (!apiKey) return [];
  const res = await fetch(`${baseUrl.replace(/\/$/, "")}/models?limit=1000`, {
    headers: { "x-api-key": apiKey, "anthropic-version": "2023-06-01" }
  });
  if (!res.ok) return [];
  const json: any = await res.json();
  const list = json?.data;
  if (!Array.isArray(list)) return [];
  return list.map((m: any) => m.id).filter(Boolean).sort();
}

async function fetchGemini(baseUrl: string, apiKey: string): Promise<string[]> {
  if (!apiKey) return [];
  const res = await fetch(`${baseUrl.replace(/\/$/, "")}/models?pageSize=1000`, {
    headers: { "x-goog-api-key": apiKey }
  });
  if (!res.ok) return [];
  const json: any = await res.json();
  const list = json?.models;
  if (!Array.isArray(list)) return [];
  return list
    .filter((m: any) => {
      const methods: string[] = m.supportedGenerationMethods ?? [];
      return methods.length === 0 || methods.includes("generateContent") || methods.includes("streamGenerateContent");
    })
    .map((m: any) => (typeof m.name === "string" ? m.name.replace(/^models\//, "") : null))
    .filter(Boolean)
    .sort();
}

export async function fetchAvailableModels(style: ApiStyle, baseUrl: string, apiKey: string): Promise<string[]> {
  try {
    switch (style) {
      case "openai":
        return await fetchOpenAiStyle(baseUrl, apiKey);
      case "anthropic":
        return await fetchAnthropic(baseUrl, apiKey);
      case "gemini":
        return await fetchGemini(baseUrl, apiKey);
    }
  } catch {
    return [];
  }
}
